src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -861,5 +861,69 @@ unsafe void M()
                     //         delegate*<C.D> d;
                     Diagnostic(ErrorCode.WRN_UnreferencedVar, "d").WithArguments("d").WithLocation(10, 24));
         }
+
+        [Fact]
+        public void FunctionPointerConstraintIntroducedBySubstitution()
+        {
+            string source = @"
+class R1<T1>
+{
+    public virtual void f<T2>() where T2 : T1 { }
+}
+class R2 : R1<delegate*<void>>
+{
+    public override void f<T2>() { }
+}
+class Program
+{
+    static void Main(string[] args)
+    {
+        R2 r = new R2();
+        r.f<int>();
+    }
+}";
+
+            var compilation = CreateFunctionPointerCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (6,7): error CS0306: The type 'delegate*<void>' may not be used as a type argument
+                // class R2 : R1<delegate*<void>>
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "R2").WithArguments("delegate*<void>").WithLocation(6, 7)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeAsThisOfExtensionMethod()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+unsafe static class C
+{
+    static void M(this delegate*<void> ptr) {}
+}");
+
+            comp.VerifyDiagnostics(
+                // (4,24): error CS1103: The first parameter of an extension method cannot be of type 'delegate*<void>'
+                //     static void M(this delegate*<void> ptr) {}
+                Diagnostic(ErrorCode.ERR_BadTypeforThis, "delegate*<void>").WithArguments("delegate*<void>").WithLocation(4, 24)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeInAnonymousType()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+unsafe static class C
+{
+    static void M(delegate*<void> ptr)
+    {
+        var a = new { Ptr = ptr };
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,23): error CS0828: Cannot assign 'delegate*<void>' to anonymous type property
+                //         var a = new { Ptr = ptr };
+                Diagnostic(ErrorCode.ERR_AnonymousTypePropertyAssignedBadValue, "Ptr = ptr").WithArguments("delegate*<void>").WithLocation(6, 23)
```

Comment:
It looks like we would report this for an array of function pointers, that feels wrong because an array can be used as a type argument. Consider adding a test that is not anonymous type specific.

Reply:
We report an error for arrays of `int*` today: [https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDZICYgNQA+AAgEwCMAsAFBEAMABEWQHQAyaAdgI7MCiAHgAcEcKFDQB7DlADc1akQDMjEvQDC9AN7V6uxsoCu0iADM4jACz0AsgApOMAFQBtALr1BMBFACUWnXqBAG4QCPQQ9AC89BxwAO5a9AAKXlBRHqn0AL5yVIFZ1FlAA=](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDZICYgNQA+AAgEwCMAsAFBEAMABEWQHQAyaAdgI7MCiAHgAcEcKFDQB7DlADc1akQDMjEvQDC9AN7V6uxsoCu0iADM4jACz0AsgApOMAFQBtALr1BMBFACUWnXqBAG4QCPQQ9AC89BxwAO5a9AAKXlBRHqn0AL5yVIFZ1FlAA=)

---
In reply to: [407753380](https://github.com/dotnet/roslyn/pull/43248#discussion_r407753380) [](ancestors = 407753380)

Reply:
I will add that to this test though.

---
In reply to: [407821869](https://github.com/dotnet/roslyn/pull/43248#discussion_r407821869) [](ancestors = 407821869,407753380)

Reply:
> Consider adding a test that is not anonymous type specific.

I assumed by this you meant not using the function pointer type as the direct type being assigned. Let me know if this wasn't what you meant.

---
In reply to: [407821909](https://github.com/dotnet/roslyn/pull/43248#discussion_r407821909) [](ancestors = 407821909,407821869,407753380)

Reply:
>> It looks like we would report this for an array of function pointers, that feels wrong because an array can be used as a type argument. Consider adding a test that is not anonymous type specific.

> I assumed by this you meant not using the function pointer type as the direct type being assigned. Let me know if this wasn't what you meant.

It appears that this test was added to test changes made in ```internal static bool IsUnsafe(this TypeSymbol type)``` function. While behavior for this particular test scenario is expected and the test is good to have, I suspect we would get the same error for an array of function pointers. I think this is unexpected, as it is unexpected for an array of regular pointers too. I assume this is due to the fact that the ```GetAnonymousTypeFieldType``` misuses the ```IsUnsafe```. I think the intent was to check for types that cannot be used as type arguments because anonymous types are generic, but it is valid to use an array of pointers as a type argument. I am going to open an issue about that. Once/if that is fixed, it is very likely ```GetAnonymousTypeFieldType``` would stop using ```IsUnsafe``` and this test would stop covering its modification. At the same time it appears that this is the only tests added to cover that. So, I am suggesting to find another scenario to cover changes in ```IsUnsafe```, a scenario that has nothing to do with anonymous types.

---
In reply to: [407827402](https://github.com/dotnet/roslyn/pull/43248#discussion_r407827402) [](ancestors = 407827402,407821909,407821869,407753380)

Reply:
Ah, understood.

---
In reply to: [408246197](https://github.com/dotnet/roslyn/pull/43248#discussion_r408246197) [](ancestors = 408246197,407827402,407821909,407821869,407753380)